### ISSUE TITLE
Books with large chapters cause extreme lag

### ISSUE THREAD

**Issue by MrXyfir** [futurepress/epub.js#714]
using v0.3.58

Testing on a book that has ~400 long paragraphs in most chapters and the browser slows dramatically, to where almost any action has a ~1 second delay. If I switch to a short chapter or another book with short chapters there's no problem. I've also tested on multiple books with large paragraphs and received the same results.

If I add `display: none` to the majority of the paragraph `p` elements then the lag instantly goes away. I was considering writing something to automatically hide and display only the appropriate elements but this screws up the pagination.

Any advice? At the current point certain books are basically unreadable. I can share a couple EPUB files I'm testing with but really any with large chapters should work.

**Comment by MrXyfir** [comment]
I've done some more testing and using `visibility: hidden` instead of `display: none` prevents the pagination from breaking and eases the lag a noticeable amount but nowhere near as much as `display: none`.

@fchasen Will there be any changes in v0.4 or beyond regarding how chapters (especially large ones) are rendered? Do you have any suggestions on how to deal with this now in v0.3? I have a few ideas but you know the code base the best, there much be an obvious solution I'm not seeing.

**Comment by MrXyfir** [comment]
Oddly, after many weeks, the lag has suddenly disappeared. I'll reopen if it returns.

**Comment by dspacejs** [comment]
Hey @MrXyfir this still seems to be an issue, do you mind reopening it? See [this issue](https://github.com/johnfactotum/foliate/issues/545) for info.

**Comment by MrXyfir** [comment]
@dspacejs No problem!

**Comment by dspacejs** [comment]
It'd be great if this can be fixed, it's been a huge pain.

@MrXyfir was it fixed when you switched to a different device or something with better specs?

**Comment by MrXyfir** [comment]
@dspacejs No, it just mysteriously disappeared. It's possible my browser updated? I haven't used this in a long time.

That being said, `display: none` worked for me at the time. If I had to solve this again I would probably try hiding the elements with `display: none` and inserting an empty `div` or something with the correct size to prevent pagination issues.

**Comment by dspacejs** [comment]
@MrXyfir I'm using another app (Foliate) which uses this library. So I'd have to fork it that, fork epub.js, and then deploy it locally as an app. That's a huge pain lol

**Comment by MrXyfir** [comment]
@dspacejs Ah, I see. Well best of luck finding a solution! Maybe you can convince the Foliate developer(s) to fork it.

**Comment by dspacejs** [comment]
@MrXyfir good idea, I'll make a PR.
